### PR TITLE
Add support for IGNORE/RESPECT NULLS clause to various window functions

### PR DESCRIPTION
**Pull request by ptkool**

### CODE REVIEW

**Comment by electrum** [comment]
@cla-bot check

File: presto-parser/src/main/java/io/prestosql/sql/parser/AstBuilder.java [hunk 1]
```diff
@@ -1558,6 +1560,7 @@ public Node visitFunctionCall(SqlBaseParser.FunctionCallContext context)
                 filter,
                 orderBy,
                 distinct,
+                ignoreNulls,
```
Comment (kasiafi) [inline presto-parser/src/main/java/io/prestosql/sql/parser/AstBuilder.java]: Please notice that the “RESPECT / IGNORE NULLS” argument is allowed by the grammar for every function call of the form `qualifiedName '(' (setQuantifier? expression (',' expression)*)? … `, which applies not only to window value functions for which it is dedicated.
Make sure that every invalid usage of RESPECT / IGNORE NULLS argument that has passed through `AstBuilder` is caught and fails with `NOT SUPPORTED`. Add appropriate checks in `ExpressionAnalyzer.visitFunctionCall()`.

File: presto-main/src/main/java/io/prestosql/operator/window/LeadFunction.java [hunk 2]
```diff
@@ -51,8 +54,19 @@ public void processRow(BlockBuilder output, int frameStart, int frameEnd, int cu
...
+            if (ignoreNulls) {
+                while (withinParition) {
+                    if (!windowIndex.isNull(valueChannel, toIntExact(valuePosition))) {
+                        break;
+                    }
+                    valuePosition++;
+                    withinParition = (valuePosition >= 0) && (valuePosition < windowIndex.size());
+                }
+            }
+
```
Comment (kasiafi) [inline presto-main/src/main/java/io/prestosql/operator/window/LeadFunction.java]: I am not sure about the semantics of this implementation. You start at `currentPosition + offset`, like it was in the previous implementation which respected nulls, and then look forward for the first non-null value. However, from reading the spec, it appears to me that you should instead start at `currentPosition` and go forward unless you encounter `offset`-th non-null value.
The same applies to `LagFunction` and `NthValueFunction`.

File: presto-docs/src/main/sphinx/functions/window.rst [hunk 3]
```diff
@@ -106,18 +114,28 @@ Value Functions
...
+    and ``x`` is null for all rows, a value of null is returned.
+
 .. function:: lead(x[, offset [, default_value]]) -> [same as input]
 
     Returns the value at ``offset`` rows after the current row in the window.
-    Offsets start at ``0``, which is the current row. The
-    offset can be any scalar expression.  The default ``offset`` is ``1``. If the
-    offset is null or larger than the window, the ``default_value`` is returned,
-    or if it is not specified ``null`` is returned.
+    Offsets start at ``0``, which is the current row. The offset can be any scalar expression.
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/window.rst]: This text looks unchanged, except for differences in the wrapping. Let's leave the wrapping alone in this diff, to make it easier to see the other changes.

File: presto-docs/src/main/sphinx/functions/window.rst [hunk 4]
```diff
@@ -106,18 +114,28 @@ Value Functions
...
+    is null for all rows, the ``default_value`` is returned.
 
 .. function:: lag(x[, offset [, default_value]]) -> [same as input]
 
     Returns the value at ``offset`` rows before the current row in the window
-    Offsets start at ``0``, which is the current row.  The
-    offset can be any scalar expression.  The default ``offset`` is ``1``. If the
-    offset is null or larger than the window, the ``default_value`` is returned,
-    or if it is not specified ``null`` is returned.
+    Offsets start at ``0``, which is the current row. The offset can be any scalar expression.
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/window.rst]: Same as above, this seems unchanged.

File: presto-parser/src/main/java/io/prestosql/sql/parser/AstBuilder.java (hunk 1 above)
Comment (electrum) [reply presto-parser/src/main/java/io/prestosql/sql/parser/AstBuilder.java]: The grammar only allows the null treatment clause for specific window functions, so I think we could change the grammar to only allow it for window functions:
```
qualifiedName '(' ... ')'
filter? (nullTreatment? over)?
```
Although we should still add the checks in the analyzer, in case it is passed in improperly built AST.

**Comment by electrum** [comment]
Please add a parser test to `TestSqlParser` that validates the AST node is built correctly.

File: presto-main/src/main/java/io/prestosql/sql/planner/plan/WindowNode.java [hunk 5]
```diff
@@ -357,6 +360,12 @@ public Frame getFrame()
             return frame;
         }
 
+        @JsonProperty
+        public boolean ignoreNulls()
```
Comment (kasiafi) [inline presto-main/src/main/java/io/prestosql/sql/planner/plan/WindowNode.java]: Rename this getter to `isIgnoreNulls`

File: presto-main/src/main/java/io/prestosql/sql/planner/plan/WindowNode.java [hunk 6]
```diff
@@ -357,6 +360,12 @@ public Frame getFrame()
...
+        {
+            return ignoreNulls;
+        }
+
         @Override
         public int hashCode()
         {
```
Comment (kasiafi) [inline presto-main/src/main/java/io/prestosql/sql/planner/plan/WindowNode.java]: Update `equals()` and `hashCode()` to take account of `ignoreNulls`

**Comment by martint** [comment]
IGNORE/RESPECT nulls is only defined for a subset of window functions in the SQL specification. We should think about how we want to deal with functions for which they are not defined. The two options I can think of are:

1. "extend' the SQL specification and support that for any window function. We need to think carefully about the implications and semantics.
2. Only allow it for functions where it makes sense. For this, we'd need additional metadata for window functions to indicate whether the clause is meaningful/relevant. The analyzer should fail the query if the cause is not allowed.

My (strong) opinion is we should do 2) for now and then think about 1) as a separate future enhancement keep this feature going.

File: presto-main/src/main/java/io/prestosql/sql/planner/plan/WindowNode.java [hunk 7]
```diff
@@ -334,12 +334,12 @@ public Function(
                 @JsonProperty("signature") Signature signature,
                 @JsonProperty("arguments") List<Expression> arguments,
                 @JsonProperty("frame") Frame frame,
-                @JsonProperty("ignoreNulls") Boolean ignoreNulls)
+                @JsonProperty("isIgnoreNulls") boolean ignoreNulls)
```
Comment (electrum) [inline presto-main/src/main/java/io/prestosql/sql/planner/plan/WindowNode.java]: This needs to be `@JsonProperty("ignoreNulls")` since `ignoreNulls` is what the name will be in the JSON (the JSON property are the logical Java Bean names -- without the `is` / `get` prefixes and starting with lowercase)

File: presto-main/src/main/java/io/prestosql/operator/window/ReflectionWindowFunctionSupplier.java [hunk 8]
```diff
@@ -29,7 +30,8 @@
 public class ReflectionWindowFunctionSupplier<T extends WindowFunction>
         extends AbstractWindowFunctionSupplier
 {
-    private final Constructor<T> constructor;
+    private Constructor<T> constructor;
```
Comment (electrum) [inline presto-main/src/main/java/io/prestosql/operator/window/ReflectionWindowFunctionSupplier.java]: Keep these as `final` so the class is immutable. The constructor can be structured so that this is possible (probably need a temporary local variable)

File: presto-main/src/main/java/io/prestosql/operator/window/ReflectionWindowFunctionSupplier.java [hunk 9]
```diff
@@ -39,11 +41,20 @@ public ReflectionWindowFunctionSupplier(String name, Type returnType, List<? ext
     public ReflectionWindowFunctionSupplier(Signature signature, Class<T> type)
     {
         super(signature, getDescription(requireNonNull(type, "type is null")));
+        canIgnoreNulls = ValueWindowFunction.class.isAssignableFrom(type);
```
Comment (electrum) [inline presto-main/src/main/java/io/prestosql/operator/window/ReflectionWindowFunctionSupplier.java]: Setting this here unconditionally isn't correct for the fallback case. The variable needs to reflect which version of the constructor is actually available.

How about adding a `ConstructorType` enum? Then the `newWindowFunction` method would be more direct and wouldn't need to re-check the signature:
```java
switch (constructorType) {
    case NO_INPUTS:
        return constructor.newInstance();
    case INPUTS:
        return constructor.newInstance(inputs);
    case INPUTS_IGNORE_NULLS:
        return constructor.newInstance(inputs, ignoreNulls);
}

File: presto-docs/src/main/sphinx/functions/window.rst [hunk 10]
```diff
@@ -121,3 +121,11 @@ Value Functions
     offset can be any scalar expression.  The default ``offset`` is ``1``. If the
     offset is null or larger than the window, the ``default_value`` is returned,
     or if it is not specified ``null`` is returned.
+
+Null Treatment
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/window.rst]: Let's remove this section header and move the paragraph to be at the top of the "Value Functions" section. So it would look like
```
Value Functions
---------------

By default, null values are respected. If ...

.. function:: first_value(x) -> [same as input]

File: presto-main/src/main/java/io/prestosql/operator/window/LagFunction.java [hunk 11]
```diff
@@ -50,9 +52,23 @@ public void processRow(BlockBuilder output, int frameStart, int frameEnd, int cu
...
 
-            long valuePosition = currentPosition - offset;
+            long valuePosition;
 
-            if ((valuePosition >= 0) && (valuePosition <= currentPosition)) {
+            if (ignoreNulls) {
+                long count = 0;
+                valuePosition = currentPosition - 1;
+                while (withinPartition(valuePosition, currentPosition)) {
+                    if (!windowIndex.isNull(valueChannel, toIntExact(valuePosition)) && ++count == offset) {
```
Comment (electrum) [inline presto-main/src/main/java/io/prestosql/operator/window/LagFunction.java]: Nit: in our code style, we don't use the increment operations inside expressions. They should be separate statements. Same applies to the other functions.

**Comment by ptkool** [comment]
@electrum @martint @kasiafi Can I get another review of this?

File: presto-parser/src/main/java/io/prestosql/sql/parser/AstBuilder.java [hunk 12]
```diff
@@ -1476,10 +1476,13 @@ public Node visitFunctionCall(SqlBaseParser.FunctionCallContext context)
 
         boolean distinct = isDistinct(context.setQuantifier());
 
+        SqlBaseParser.NullTreatmentContext nullTreatment = context.nullTreatment();
+
         if (name.toString().equalsIgnoreCase("if")) {
             check(context.expression().size() == 2 || context.expression().size() == 3, "Invalid number of arguments for 'if' function", context);
             check(!window.isPresent(), "OVER clause not valid for 'if' function", context);
             check(!distinct, "DISTINCT not valid for 'if' function", context);
+            check(nullTreatment == null, "Null treatment clause not valid for 'if' function", context);
```
Comment (electrum) [inline presto-parser/src/main/java/io/prestosql/sql/parser/AstBuilder.java]: I don't think these checks are needed anymore, since we restrict `nullTreatment` to be paired with an `over` clause in the grammar.

File: presto-parser/src/main/java/io/prestosql/sql/tree/FunctionCall.java [hunk 13]
```diff
@@ -48,6 +49,7 @@ public FunctionCall(
             Optional<Expression> filter,
             Optional<OrderBy> orderBy,
             boolean distinct,
+            boolean ignoreNulls,
             List<Expression> arguments)
```
Comment (kasiafi) [inline presto-parser/src/main/java/io/prestosql/sql/tree/FunctionCall.java]: Please update `equals()` and `hashCode()` to take account of the new field `ignoreNulls`.

**Comment by kasiafi** [comment]
Could you please add a test with two similar calls varying in null treatment, to make sure the issue mentioned by @willmo83 is fixed?

**Comment by electrum** [comment]
@ptkool Thanks for updating. The fixes look good. Please add a test for the issue found by @willmo83, and squash everything into a single commit. Then I can merge after Travis passes.

File: presto-main/src/main/java/io/prestosql/operator/window/ReflectionWindowFunctionSupplier.java [hunk 14]
```diff
@@ -29,7 +31,16 @@
 public class ReflectionWindowFunctionSupplier<T extends WindowFunction>
         extends AbstractWindowFunctionSupplier
 {
+    @SuppressWarnings("unchecked")
```
Comment (electrum) [inline presto-main/src/main/java/io/prestosql/operator/window/ReflectionWindowFunctionSupplier.java]: Remove this `@SuppressWarnings` as it's not needed

**Comment by electrum** [comment]
The new test for the issue of multiple instances of a function should go in `io.prestosql.tests.AbstractTestWindowQueries`, as that runs a full query and can have multiple window function invocations.

**Comment by electrum** [comment]
Merged, thanks!

**Comment by electrum** [comment]
Merged as 4fba34f3d94291d5411a5c24bd47fa0213898e51
